src/combinatorics/colex.lean

```diff
@@ -0,0 +1,291 @@
+  rintros ⟨k₁, k₁z, notinA, inB⟩ ⟨k₂, k₂z, notinB, inC⟩,
+  cases lt_or_gt_of_ne (ne_of_mem_of_not_mem inB notinB),
+  { refine ⟨k₂, _, by rwa k₁z h, inC⟩,
+    intros x hx,
+    rw ← k₂z hx,
+    apply k₁z (trans h hx) },
+  { refine ⟨k₁, _, notinA, by rwa ← k₂z h⟩,
+    intros x hx,
+    rw k₁z hx,
+    apply k₂z (trans h hx) }
+end
+
+@[trans]
+lemma le_trans [linear_order α] (a b c : finset.colex α) :
+  a ≤ b → b ≤ c → a ≤ c :=
+λ AB BC, AB.elim (λ k, BC.elim (λ t, or.inl (lt_trans k t)) (λ t, t ▸ AB)) (λ k, k.symm ▸ BC)
+
+instance [linear_order α] : is_trans (finset.colex α) (<) := ⟨λ _ _ _, colex.lt_trans⟩
+instance [linear_order α] : is_asymm (finset.colex α) (<) := by apply_instance
+
+instance [linear_order α] : is_strict_order (finset.colex α) (<) := {}
+
+lemma lt_trichotomy [linear_order α] (A B : finset.colex α) :
+  A < B ∨ A = B ∨ B < A :=
+begin
+  by_cases h₁ : (A = B),
+  { right,
+    left,
+    assumption  },
+  rcases (exists_max_image (A \ B ∪ B \ A) id _) with ⟨k, hk, z⟩,
+  { simp only [mem_union, mem_sdiff] at hk,
+    cases hk,
+    { right,
+      right,
+      refine ⟨k, λ t th, _, hk.2, hk.1⟩,
+      specialize z t,
+      by_contra h₂,
+      simp only [mem_union, mem_sdiff, id.def] at z,
+      rw [not_iff, iff_iff_and_or_not_and_not, not_not, and_comm] at h₂,
+      apply not_le_of_lt th (z h₂) },
+    { left,
+      refine ⟨k, λ t th, _, hk.2, hk.1⟩,
+      specialize z t,
+      by_contra h₃,
+      simp only [mem_union, mem_sdiff, id.def] at z,
+      rw [not_iff, iff_iff_and_or_not_and_not, not_not, and_comm, or_comm] at h₃,
+      apply not_le_of_lt th (z h₃) }, },
+  rw nonempty_iff_ne_empty,
+  intro a,
+  simp only [union_eq_empty_iff, sdiff_eq_empty_iff_subset] at a,
+  apply h₁ (subset.antisymm a.1 a.2)
+end
+
+instance [linear_order α] : is_trichotomous (finset.colex α) (<) := ⟨lt_trichotomy⟩
+
+-- It should be possible to do this computably but it doesn't seem to make any difference for now.
+noncomputable instance [linear_order α] : linear_order (finset.colex α) :=
+{ le_refl := λ A, or.inr rfl,
+  le_trans := le_trans,
+  le_antisymm := λ A B AB BA, AB.elim (λ k, BA.elim (λ t, (asymm k t).elim) (λ t, t.symm)) id,
+  le_total := λ A B,
+          (lt_trichotomy A B).elim3 (or.inl ∘ or.inl) (or.inl ∘ or.inr) (or.inr ∘ or.inl),
+  decidable_le := classical.dec_rel _,
+  ..finset.colex.has_le }
+
+instance [linear_order α] : is_incomp_trans (finset.colex α) (<) :=
+begin
+  constructor,
+  rintros A B C ⟨nAB, nBA⟩ ⟨nBC, nCB⟩,
+  have: A = B := ((lt_trichotomy A B).resolve_left nAB).resolve_right nBA,
+  have: B = C := ((lt_trichotomy B C).resolve_left nBC).resolve_right nCB,
+  rw [‹A = B›, ‹B = C›, and_self],
+  apply irrefl
+end
+
+instance [linear_order α] : is_strict_weak_order (finset.colex α) (<) := {}
+instance [linear_order α] : is_strict_total_order (finset.colex α) (<) := {}
+
+/-- Colex is an extension of the base ordering on α. -/
+lemma singleton_lt_iff_lt [linear_order α] {r s : α} :
```

Comment:
Maybe this should be split into two parts, with one singleton on each side: `{r} <ᶜ s` iff `s` contains an element greater than `r` and `s <ᶜ {r}` iff all elements of `s` are less than `r`. (If I recall the properties of `colex` correctly.)

Reply:
I think this is a good idea, but this lemma as it is should still be there as a consequence

Reply:
I don't think the first statement is true - a counterexample is ``{2} <c {1, 2}``